README.md
# Subscribe by Email

**INACTIVE NOTICE: This plugin is unsupported by WPMUDEV, we've published it here for those technical types who might want to fork and maintain it for their needs.**

## Translations

Translation files can be found at https://github.com/wpmudev/translations


## Subscribe by Email allows your audience to easily subscribe to personalized, clean, regular HTML digest emails.

 

![Use the color picker for a simple color match.](https://premium.wpmudev.org/wp-content/uploads/2010/04/color-pick-735x470.jpg)

 Use the color picker for a simple color match.

### Polished Look and Feel

Tailor your feed to fit your aesthetic with a few styling tweaks. Add a custom logo, pick colors, give it a title, a subject, configure image controls and set footer content. Customize your feed to only send updates for specific categories and post types.   

![A built-in Follow Box makes it easy to capture new subscribers.](https://premium.wpmudev.org/wp-content/uploads/2010/04/email-subscribe-735x470.jpg)

 The built-in Follow Box makes it easy to capture new subscribers.

### Subscribe and Sign-Up

The built-in Follow Box slide-in and Subscribe widget make it super simple for users to sign-up for your email updates. Manage a ton of users and save a bunch of time with the bulk subscribe importer. 

### On Your Schedule

Take control of when and how often your emails are sent out. Configure to send immediately or send a daily or weekly digest. Plus, Subscribe by Email includes throttle controls so you don't have to worry about over-working your server or spam blacklisting.  

![Layout your design, get a live preview and send a test for a polished end-user experience.](https://premium.wpmudev.org/wp-content/uploads/2010/04/test-email-735x470.jpg)

 Layout your design, get a live preview and send a test for a polished end-user experience.

### Double Opt-In

Your subscribers will be comfortable knowing, with double opt-in they can manage and stop their subscriptions at any time. Give your users control with automatically included, clear simple unsubscribe links on every email. 

## Usage

### To Get Started:

Start by reading the [Installing Plugins](https://premium.wpmudev.org/manuals/wpmu-manual-2/installing-regular-plugins-on-wpmu/) section in our comprehensive [WordPress and WordPress Multisite Manual](https://premium.wpmudev.org/wpmu-manual/) if you are new to WordPress.

### Configuring the Settings

Once installed and activated, you will see a new Subscriptions menu item in your site's dashboard. Go ahead and click on the Settings sub-menu now. [

![subscribe-by-email-2300-menu](https://premium.wpmudev.org/wp-content/uploads/2010/04/subscribe-by-email-2300-menu.png)

](https://premium.wpmudev.org/wp-content/uploads/2010/04/subscribe-by-email-2300-menu.png)

 Under the _General Settings_ tab, you can enter your basic information, control how emails are sent and for which post types, as well as set a page for user subscription management. 

![1\. Enter your header notification text. 2\. Enter the send-from email address. 3\. Enter the email subject line. 4\. Select how often emails are sent. 5\. Enter the number of emails to send every 30 minutes. 6\. Select the role to notify of subscription changes. 7\. Select the page to use for user subscription management. 8\. Toggle a follow button on/off.](https://premium.wpmudev.org/wp-content/uploads/2010/04/subscribe-by-email-2430-settings-gen.png)

 1\. Enter your header notification text.  
2\. Enter the send-from email address.  
3\. Enter the email subject line.  
4\. Select how often emails are sent.  
5\. Enter the number of emails to send every 30 minutes.  
6\. Select the role to notify of subscription changes.  
7\. Select the page to use for user subscription management.  
8\. Toggle a follow button on/off.

 1\. The _Notification From Sender_ is what is displayed to subscribers in their email client before they open your email. Be sure to use only a few words, and that it describes your mailing well so that they don't spam or delete your mail. This also appears in the header of the email template once opened. 2\. The _Notification From Email_ is the sent-from address that is displayed in the email client header when your subscribers open your mail. 3\. The _Mail subject_ appears in the email subject line of your subscribers email clients, and is displayed at the top of the email content. 4\. The _Email Frequency_ setting has 4 options to select from:

*   _Immediately when a new post is published_ will fire off an email to your subscribers every time a selected post type is published on your blog. You may want to select a different option if you have a particularly active blog.
*   _Send a weekly digest with all posts from the previous week_ will do just that. Excerpts from all selected post types published in the previous week will be included in the digest email. Selecting this option will open a new setting where you can select the day of the week to send your weekly digest.
*   _Send a daily digest with all posts from the previous 24 hours_ is ideal for active sites where subscribers really want to be kept in the loop. An additional setting will appear so you can enter the time of day to send the digest. Note that this depends on the time you have set for your site under "Settings" > "General".
*   _Never_ will stop all emails from being sent.

5\. _Mail batches_ enables you to throttle delivery to large mailing lists if you experience timeouts or other server issues. 6\. _Get notifications_ allows you to select a user role to receive notification of new subscribers or cancelled subscriptions. 7\. The _Subscribers Management Page_ should be a page you create specifically for this use. This is the page your subscribers will land on when they click the "manage your subscriptions" link in the footer of any email they receive from you. Your users can easily unsubscribe from automatic notifications of any post type, or subscribe to a new one, simply by selecting from the list. You can include any content you like on the Subscription page; the subscription management part will be added automatically after your content. Note that the form depends on the user's unique subscription key, so it will only display when visited by clicking the "manage" link in an email. 8\. Check _Display a follow button?_ if you want the popup subscribe form to be available for your users to subscribe to your blog instantly. 

![Click the Follow button to pop open the subscribe form](https://premium.wpmudev.org/wp-content/uploads/2010/04/subscribe-by-email-2430-follow.png)

 Click the Follow button to pop open the subscribe form

### Selecting Email Contents

The Contents tab will list all the post types active on your site. You can select which post types and categories to include in emails sent to your subscribers. 

![subscribe-by-email-2430-settings-contents](https://premium.wpmudev.org/wp-content/uploads/2010/04/subscribe-by-email-2430-settings-contents.png)

 The post types you set here will also be available for selection by your subscribers on the Subscribers Management Page. This allows them to opt out of receiving notification of selected post types. 

![subscribe-by-email-2430-subscription-management](https://premium.wpmudev.org/wp-content/uploads/2010/04/subscribe-by-email-2430-subscription-management.png)

### Styling Your Emails

Next up: getting emails looking the way you want them to. The settings under the _Mail Template_ tab enable you to brand your emails with your preferred look & feel. 

![1\. Upload and style your logo. 2\. Set your email header options. 3\. Set your email footer options. 4\. Enter the confirmation email text. 5\. Set additional options.](https://premium.wpmudev.org/wp-content/uploads/2010/04/subscribe-by-email-2430-settings-style.png)

 1\. Upload and style your logo.  
2\. Set your email header options.  
3\. Set your email footer options.  
4\. Enter the confirmation email text.  
5\. Set additional options.

 1\. The _Logo for notifications_ that you upload here is the image that will display in the header of all emails sent by this plugin. You can also set the maximum width for the logo in pixels. 2\. The _Header color_ and _Header text color_ settings are self-explanatory. The _Subtitle text_ appears in the email header. If checked, _Show From Sender_ will display the Notification From Sender you entered in the General Settings. 3\. The _Footer text_ setting is where you can include any additional information you want at the very bottom of every email. This can be especially useful for contact info, etc. Here also, any HTML will render as text in the email. 4\. The _Subscribe Email Content_ setting allows you to customize the content of the email users receive when they subscribe. There is a link in that email that they must click in order to confirm their subscription. 5\. Checking the _Show featured images_ checkbox will ensure that the plugin attempts to include the featured image from each post in emails. Note that this may not function perfectly with every theme.

### Live Preview

At any time during the setup, you can preview how your emails will look by clicking the _Refresh Changes_ button. Once you're done with your settings, click _Save Changes_ at the bottom of the screen. If you want to start over from scratch, simply click the _Restore template to default_ link. Note that this will only clear your style settings, and all your general settings will remain intact. Here's what the email preview would look like using the example settings from our screenshots above: 

![subscribe-by-email-2430-settings-preview](https://premium.wpmudev.org/wp-content/uploads/2010/04/subscribe-by-email-2430-settings-preview.png)

### Email Log

Every time an email is sent by this plugin, it is logged and you can view their details on the _Sent Emails_ page.

### Managing Subscribers

You can instantly see details about your subscribers by clicking the _Subscribers_ sub-menu item. The subscribers page will display a sortable table of all users who have subscribed to your blog. 

![subscribe-by-email-2430-subscribers](https://premium.wpmudev.org/wp-content/uploads/2010/04/subscribe-by-email-2430-subscribers.png)

 You will see each subscriber's email address, the date they subscribed, whether or not they have clicked the link in their confirmation email, as well as the type of subscription. Confirmation email? Yes, this plugin has a double opt-in feature by default. New subscribers will receive an email containing a link they must click to confirm their subscription. If not clicked within 7 days, their data is deleted from the system. If you ever need to re-send a confirmation email to a subscriber within that 7-day window, simply hover your mouse pointer over the subscriber's email address and click the _Resend confirmation mail_ link. You can also cancel a user's subscription in the same manner. If you need to cancel multiple subscriptions, tick the checkboxes and select the option from the _Bulk Actions_ dropdown. The information in the Subscription Type column simply refers to how the users subscribed. Let's take a look at how your blog readers can be subscribed.

### Creating Subscriptions

There are 4 ways that subscriptions can be created. Let's go through all of them.

##### Subscribe by Email Widget

Your blog readers can subscribe via the Subscribe by Email widget. It is very simple. Go to _Appearance > Widgets_ and drag the _Subscribe by Email_ widget to your sidebar. You can customize the title, the widget text, and even the button. Here's what the widget looks like (settings on the left, front-end widget appearance on the right): 

![subscribe-by-email-2430-widget](https://premium.wpmudev.org/wp-content/uploads/2010/04/subscribe-by-email-2430-widget.png)

##### Subscribe by Email Shortcode

You can also use the following shortcode to display a subscribe form on any page, post or other post type: `[subscribe-by-email-form]` You can easily add that using the Subscribe by Email icon in the visual editor as well. 

![Subscribe by Email - visual editor icon](https://premium.wpmudev.org/wp-content/uploads/2010/04/Subscribe-by-Email-2.8.3-visual-editor-icon.png)

##### Follow button

With the "_Display a follow button?_" setting enabled, users will see a Follow button in the site's front-end. They can click the button to open a small form which they can use to subscribe. 

![Subscribe by Email - Follow button](https://premium.wpmudev.org/wp-content/uploads/2010/04/Subscribe-by-Email-2.8.3-Follow-button.png)

##### Manually adding subscribers

You can also manually add subscribers, simply click _Add Subscribers_ in your Subscriptions menu. [

![Add new user](https://premium.wpmudev.org/wp-content/uploads/2010/04/add_new_user.jpg)

](https://premium.wpmudev.org/wp-content/uploads/2010/04/add_new_user.jpg)

 You can add subscribers one at a time simply by entering an email address and clicking the _Subscribe_ button. Or you can import a CSV file of email address to add them in bulk. Your CSV file should contain only email addresses with no header, so it would look something like this: 

![subscribe-by-email-csv-2110](https://premium.wpmudev.org/wp-content/uploads/2010/04/subscribe-by-email-csv-2110.png)

 Simply click the _Browse_ button, select the CSV file on your computer and click _Import_.

#### Canceling Subscriptions

You saw above that you, as admin, can cancel subscriptions on the _Subscribers_ page. Your subscribers can also opt out of their subscription at any time simply by clicking the link in the footer of any email they receive from you. 

![subscribe-by-email-cancel-2110](https://premium.wpmudev.org/wp-content/uploads/2010/04/subscribe-by-email-cancel-2110.png)

 And that's it for all the plugin settings. Now it's up to you to write the best stuff on the web, and get it out to your new subscribers! Happy blogging!


changelog.txt
Plugin Name: Subscribe by Email
Author: WPMU DEV
Requires at least: 3.5
Tested up to: 4.7

Change Log:
----------------------------------------------------------------------
----------------------------------------------------------------------

3.6.0 - 2018-07-17
----------------------------------------------------------------------
- Improved: Accessibility
- Added: reCAPTCHA

3.5.4 - 2017-02-21
----------------------------------------------------------------------
- Fixed: XSS Vulnerability
- Some minor improvements

3.5.2 2015-02-02
----------------------------------------------------------------------
- Add: New filter to filter widget instance on validation

3.5.1 2015-11-17
----------------------------------------------------------------------
- Fixed: From email in a subfolder multisite was set to 'no-reply@'

3.5 2015-11-15
----------------------------------------------------------------------
- Enhance: Fixed some translations.
- Enhance: Improved handler to avoid duplicated digests.
- Enhance: Fixed MySQL 5.6 incompatibility
- Fixed: Typo in email template.
- New hooks.

3.4.4 2015/09/15
----------------------------------------------------------------------
- Fixed: Fatal error when upgrading from a very old version

3.4.3 2015/08/22
----------------------------------------------------------------------
- Fixed: When deleting a site in a multisite the queue table was deleted.

3.4.2 2015/08/11
----------------------------------------------------------------------
- Updated widget to WP 4.3 constructor

3.4.1 2015/07/22
----------------------------------------------------------------------
- Do not send Metabox was not working properly for immediately sendings.

3.4 2015/07/15
----------------------------------------------------------------------
- New: Added meta box to avoid sending single posts
- Enhanced: Added custom capability

3.3 2015/07/07
----------------------------------------------------------------------
- New: Preview template in real time.
- New: auto opt in attribute in shortcode.

3.2.1 2015/06/08
----------------------------------------------------------------------
- Fixed: Full post/excerpt content were not being rendered in digests.

3.2 2015/06/01
----------------------------------------------------------------------
- Enhancement: Email templates can now be overriden with custom templates in theme folder.
- Fixed typos.
- Fixed several warnings.

3.1.2 2015/04/30
----------------------------------------------------------------------
- Added script to scroll down to the shortcode form if there have been errors while processing it.

3.1.1 2015/04/30
----------------------------------------------------------------------
- Fixed: Page title when a user subscribes/unsubscribes

3.1 2015/04/20
----------------------------------------------------------------------
- Enhancement: Removed old legacy code.
- Enhancement: Plugin is now faster loading.
- Fixed: Post types with no taxonomies were not being sent.
- Added unit tests folder

3.0.3 2015/04/10
----------------------------------------------------------------------
- Enhancement: Additional check to avoid duplicated emails.
- Fixed: TinyMCE Editor button now updated for WordPress 4.1

3.0.2 - 2015/03/16
----------------------------------------------------------------------
- Enhancement: When an email fails, error is now displayed in Emails Logs
- Other minor fixes

3.0.1 - 2015/01/31
----------------------------------------------------------------------
- Fixed: Email address missing in notification mail

3.0 - 2015/01/27
----------------------------------------------------------------------
- Enhancement: New hooks.
- Enhancement: Administrators notifications are now sent once the user is confirmed in order to avoid spam.
- Fixed: Sending emails by taxonomies was not working right.
- Fixed: When activating in a subsite inside a Network, from email was set wrong.
- Fixed: Character encoding in subscribe/unsubscribe notices.

2.9.3 - 2015/01/12
----------------------------------------------------------------------
- Fixed: Fatal error when deleting custom fields

2.9.2 - 2014/11/27
----------------------------------------------------------------------
- Fixing: When creating a new campaign, the max subscriber ID to send was set wrong
- Enhancement: Added actions to finish campaigns manually

2.9.1 - 2014/11/26
----------------------------------------------------------------------
- Fixed: Queue table was not created in some cases.
- Fixed: Sites in multisite were not sending their own queue but other sites queue
- Other minor fixes

2.9 - 2014/11/03
----------------------------------------------------------------------
- Enhancement: Added queue screen that allows more control over emails that are going to be sent.
- Enhancement: Trigger next batch of emails with a button.
- Enhancement: Many performance improvements
- Enhancement: UI/Icons Improvements
- Enhancement: Added more filters.
- Fixed: Settings needed to be saved twice when the plugin was activated.
- Other minor fixes.

2.8.3 - 2014/08/29
----------------------------------------------------------------------
- Preview email was only sent if the email was a subscriber.

2.8.2 - 2014/08/08
----------------------------------------------------------------------
Fixed: Added trailing slash to Unsubscribe-List header
Enhanced: Logo uploader updated to WP 3.5+ Media Uploader
Enhanced: Added more hooks

2.8.1.2 - 2014/08/06
----------------------------------------------------------------------
Fixed: Fatal error when downloading CSV
Fixed: warning on manage subscriptions page
Fixed: Manage subscriptions page link might appear in emails even if there wasn't any selected
Fixed: Removing text inserted by mistake in network settings

2.8.1.1 - 2014/07/08

inc/settings.php
<?php

class Incsub_Subscribe_By_Email_Settings_Handler {

	static $instance;

	// Settings of the plugin
	private $settings = array();
	private $default_settings;

	// Settings properties
	private $frequency;
	private $time;
	private $day_of_week;
	private $confirmation_flag;
	private $post_types;
	private $taxonomies;
	private $follow_button_schemas;

	// Settings slug name
	private $settings_slug = 'incsub_sbe_settings';
	private $settings_network_slug = 'incsub_sbe_network_settings';

	public function __construct() {
		$this->set_settings();
		add_action( 'init', array( &$this, 'set_settings' ) );
	}


	/**
	 * Singleton Pattern
	 * 
	 * Gets the instance of the class
	 */
	public static function get_instance() {
		if ( empty( self::$instance ) )
			self::$instance = new self();
		return self::$instance;
	}

	/**
	 * Set the settings for the plugin
	 */
	public function set_settings() {

		global $wp_locale;

		$this->settings_slug = 'incsub_sbe_settings';

		// The user can choose between these options
		$this->frequency = array(
			'inmediately' 	=> __( 'Immediately when a new post is published', INCSUB_SBE_LANG_DOMAIN ),
			'weekly'		=> __( 'Send a weekly digest with all posts from the previous week', INCSUB_SBE_LANG_DOMAIN ),
			'daily'			=> __( 'Send a daily digest with all posts from the previous 24 hours', INCSUB_SBE_LANG_DOMAIN ),
			'never'			=> __( 'Never', INCSUB_SBE_LANG_DOMAIN )
		);

		$this->time = array();
		for ( $i = 0; $i < 24; $i++ ) {
			$this->time[$i] = str_pad( $i, 2, 0, STR_PAD_LEFT ) . ':00';
		}

		$this->day_of_week = $wp_locale->weekday;

		$this->confirmation_flag = array(
			0 => __( 'Awaiting confirmation'),
			1 => __( 'Email confirmed')
		);

		$current_settings = $this->get_blog_settings();


		if ( is_multisite() ) {
			$current_network_settings = $this->get_network_settings();
			$current_settings = array_merge( $current_settings, $current_network_settings );
		}

		$this->settings = wp_parse_args( $current_settings, $this->get_default_settings() );

		// Get all post types
		$args = array(
		  'publicly_queryable'   => true,
		); 
		$this->post_types = get_post_types( $args, 'object' );
		unset( $this->post_types['attachment'] );
		$this->post_types = apply_filters( 'sbe_get_post_types', $this->post_types );

		// Get those taxonomies that are hierachical
		$this->taxonomies = array();
		foreach ( $this->post_types as $post_slug => $post_type ) {
			$post_type_taxonomies = get_object_taxonomies( $post_slug );

			if ( ! empty( $post_type_taxonomies ) ) {
				foreach ( $post_type_taxonomies as $taxonomy_slug ) {
					$taxonomy = get_taxonomy( $taxonomy_slug );

					if ( $taxonomy->hierarchical ) {
						$this->taxonomies[ $post_slug ][ $taxonomy_slug ] = $taxonomy;
					}
				}
				
			}

		}

		$this->extra_field_types = array( 
			'text' => array(
				'name' => __( 'Text field', INCSUB_SBE_LANG_DOMAIN ),
				'slug' => 'text'
			), 
			'checkbox' => array(
				'name' => __( 'Checkbox', INCSUB_SBE_LANG_DOMAIN ),
				'slug' => 'checkbox'
			)
		);

		$this->follow_button_schemas = array( 
			'dark' => array(
				'label' => __( 'Dark', INCSUB_SBE_LANG_DOMAIN ),
				'slug' => 'dark'
			),
			'light' => array(
				'label' => __( 'Light', INCSUB_SBE_LANG_DOMAIN ),
				'slug' => 'light' 
			) 
		);

		$this->follow_button_positions = array( 
			'bottom' => array(
				'label' => __( 'Bottom', INCSUB_SBE_LANG_DOMAIN ),
				'slug' => 'bottom'
			),
			'left' => array(
				'label' => __( 'Left', INCSUB_SBE_LANG_DOMAIN ),
				'slug' => 'left' 
			),
			'right' => array(
				'label' => __( 'Right', INCSUB_SBE_LANG_DOMAIN ),
				'slug' => 'right' 
			) 
		);

	}

	public function get_network_settings() {
		$defaults = $this->get_default_network_settings();
		$default_keys = array_keys( $defaults );
		$current_settings = get_site_option( $this->settings_network_slug, $defaults );
		$network_settings = array();
		foreach ( $current_settings as $current_setting => $value ) {
			if ( in_array( $current_setting, $default_keys ) )
				$network_settings[ $current_setting ] = $value;
		}
		return wp_parse_args( $network_settings, $defaults );
	}

	public function get_blog_settings() {
		$defaults = $this->get_default_blog_settings();

		if ( ! is_multisite() )
			$defaults = array_merge( $defaults, $this->get_default_network_settings() );
		
		$default_keys = array_keys( $defaults );
		$current_settings = get_option( $this->settings_slug, $defaults );
		$blog_settings = array();

		if ( ! is_array( $current_settings ) )
			$current_settings = $defaults;
		
		foreach ( $current_settings as $current_setting => $value ) {
			if ( in_array( $current_setting, $default_keys ) )
				$blog_settings[ $current_setting ] = $value;
		}

		return wp_parse_args( $blog_settings, $defaults );
	}

	public function get_settings() {
		if ( empty( $this->settings ) )
			$this->set_settings();

		return $this->settings;
	}

	public function update_settings( $settings ) {

		if ( is_multisite() ) {
			$network_settings = $this->get_network_settings();
			$blog_settings = $this->get_blog_settings();

			foreach ( $settings as $setting => $value ) {
				if ( array_key_exists( $setting, $network_settings ) )
					$network_settings[ $setting ] = $value;
				elseif ( array_key_exists( $setting, $blog_settings ) )
					$blog_settings[ $setting ] = $value;
			}

			update_site_option( $this->settings_network_slug, $network_settings );
			update_option( $this->settings_slug, $blog_settings );

		}
		else {
			update_option( $this->settings_slug, $settings );
		}

		$this->set_settings();
	}

	public function get_default_settings() {
		return array_merge( $this->get_default_blog_settings(), $this->get_default_network_settings() );
	}

	public function get_default_blog_settings() {

		global $current_site;

		$subscribe_email_content = _x( 'Howdy.', 'Confirmation email sent to subscribers.', INCSUB_SBE_LANG_DOMAIN ) . "\r\n\r\n";
		$subscribe_email_content .= __( 'You recently signed up to be notified of new posts on my blog. This means', INCSUB_SBE_LANG_DOMAIN ) . "\r\n";
		$subscribe_email_content .= __( 'once you confirm below, you will receive an email when posts are published.', INCSUB_SBE_LANG_DOMAIN ) . "\r\n";
		$subscribe_email_content .= __( 'To activate, click confirm below. If you believe this is an error, ignore this message and nothing more will happen.', INCSUB_SBE_LANG_DOMAIN ) . "\r\n";

		$defaults = array(
			'auto-subscribe' => false,
			'subscribe_new_users' => false,
			'from_sender' => html_entity_decode( get_bloginfo( 'name' ), ENT_QUOTES ),
			'subject' => html_entity_decode( get_bloginfo( 'name' ) . __( ': New post' ), ENT_QUOTES ),
			'frequency' => 'inmediately',
			'time' => 0,
			'day_of_week' => 0,
			'manage_subs_page' => 0,
			'get_notifications' => false,
			'get_notifications_role' => 'administrator',

			'follow_button' => false,
			'follow_button_schema' => 'dark',
			'follow_button_position' => 'bottom',

			'post_types' => array( 'post' ),
			'taxonomies' => array( 'post' => array( 'category' => array( 'all' ) ) ),

			'logo' => '',
			'logo_width' => 200,
			'featured_image' => false,
			'header_text' => '',
			'show_blog_name' => true,
			'footer_text' => '',
			'header_color' => '#66aec2',
			'header_text_color' => '#000000',
			
			'send_full_post' => false,
			'subscribe_email_content' => $subscribe_email_content,

			'extra_fields' => array(),

			'recaptcha_type' => 'no',
			'recaptcha_public_key' => '',
			'recaptcha_secret_key' => ''
		);

		return apply_filters( 'sbe_blog_default_settings', $defaults );
	}

	function sanitize_template_settings( $new_settings ) {
		$settings = incsub_sbe_get_settings();

		foreach ( $settings as $setting_name => $setting_value ) {
			if ( isset( $new_settings[ $setting_name ] ) ) {
				switch ( $setting_name ) {
					case 'logo': {
						if ( empty( $new_settings[ $setting_name ] ) )
							$settings[ $setting_name ] = '';
						else
							$settings[ $setting_name ] = esc_url_raw( $new_settings[ $setting_name ] );
						break;
					}
					case 'logo_width': {
						if ( isset( $new_settings[ $setting_name ] ) && is_numeric( $new_settings[ $setting_name ] ) ) {
							$settings[ $setting_name ] = absint( $new_settings[ $setting_name ] );
						}
						break;
					}
					case 'send_full_post':
					case 'show_blog_name':
					case 'featured_image': {
						$settings[ $setting_name ] = (bool)$new_settings[ $setting_name ];
						break;
					}
					case 'header_color':
					case 'header_text_color': {
						if ( preg_match('|^#([A-Fa-f0-9]{3}){1,2}$|', $new_settings[ $setting_name ] ) )
							$settings[ $setting_name ] = $new_settings[ $setting_name ];
						break;
					}
					case 'header_text': 
					case 'footer_text': 
					case 'subscribe_email_content':  {
						$settings[ $setting_name ] = wp_kses( $new_settings[ $setting_name ], wp_kses_allowed_html() );
					}
				}
			}
		}

		return $settings;
	}

	function get_default_network_settings() {
		if ( is_multisite() && is_subdomain_install() ) {
			$current_site = get_current_site();
			$blog_details = get_blog_details( $current_site->blog_id );
			$base_domain = $blog_details->domain;
		}
		elseif ( is_multisite() && ! is_subdomain_install() ) {
			$current_site = get_current_site();
			$blog_details = get_blog_details( $current_site->blog_id );
			$base_domain = $blog_details->domain;
			$base_domain = preg_replace( '/^\//', '', $base_domain );
			$base_domain = preg_replace( '/\/$/', '', $base_domain );
			$base_domain = str_replace( '/', '.', $base_domain );
		}
		else {
			$base_domain = get_bloginfo('wpurl');
			$base_domain = str_replace( 'http://', '', $base_domain );
			$base_domain = preg_replace( '/^\//', '', $base_domain );
			$base_domain = preg_replace( '/\/$/', '', $base_domain );
			$base_domain = str_replace( '/', '.', $base_domain );
			$base_domain = str_replace( 'www.', '', $base_domain );
		}

		return array(
			'from_email' => 'no-reply@' . $base_domain,
			'keep_logs_for' => 31,
			'mails_batch_size' => 80
		);
	}

	/**
	 * Get the settings slug name
	 * 
	 * @return String
	 */
	public function get_settings_slug() {
		return $this->settings_slug;
	}

	public function get_frequency() {
		return $this->frequency;
	}

	public function get_time() {
		return $this->time;
	}

	public function get_day_of_week() {
		return $this->day_of_week;
	}

	public function get_confirmation_flag() {
		return $this->confirmation_flag;
	}

	public function get_post_types() {
		return $this->post_types;
	}

	public function get_taxonomies() {
		return $this->taxonomies;
	}

	public function get_taxonomies_by_post_type( $post_type_slug ) {
		return isset( $this->taxonomies[ $post_type_slug ] ) ? $this->taxonomies[ $post_type_slug ] : array();
	}

	public function get_selected_taxonomies_by_post_type( $post_type_slug ) {
		$taxonomies = array();
		if ( isset( $this->settings['taxonomies'][ $post_type_slug ] ) && in_array( 'all', $this->settings['taxonomies'][ $post_type_slug ] ) )
			return $this->get_taxonomies_by_post_type( $post_type_slug );
	}

	public function get_extra_field_types() {
		return $this->extra_field_types;
	}

	public function get_follow_button_schemas() {
		return $this->follow_button_schemas;
	}

	public function get_follow_button_positions() {
		return $this->follow_button_positions;
	}

}